### PR TITLE
snat match use acl

### PR DESCRIPTION
**Pull request by donghaobo**
snat match use acl

### CODE REVIEW

File: src/ipvs/ip_vs_service_match_acl.c [hunk 1]
```diff
@@ -0,0 +1,929 @@
...
+        dp_vs_svc_match_acl4[cid] = NULL;
+        dp_vs_svc_match_acl6[cid] = NULL;
+    }
+    stop_build = 0;
+    pthread_t tid;
+    if (pthread_create(&tid, NULL, dpvs_svc_match_build_loop, NULL)) {
+        RTE_LOG(ERR, IPVS, "fail to create dpvs_svc_match_build_loop thread\n");
+        dp_vs_svc_match_term();
+        return EDPVS_SYSCALL;
+    }
```
Comment (ywc689) [inline src/ipvs/ip_vs_service_match_acl.c:877]: DPVS is based on a poll mode driver, with each thread pinned to an exclusive cpu, and extra threading is not encouraged for performance considerations.

File: src/ipvs/ip_vs_service.c [hunk 2]
```diff
@@ -36,21 +37,46 @@
...
+#include "ipvs/service_match_acl.h"
 
-static rte_atomic16_t dp_vs_num_services[DPVS_MAX_LCORE];
+static rte_atomic32_t dp_vs_num_services[DPVS_MAX_LCORE];
 
 /**
  * hash table for svc
  */
-#define DP_VS_SVC_TAB_BITS 8
+#define DP_VS_SVC_TAB_BITS 14
```
[the remark this answers is not in the document]
Comment (donghaobo) [reply src/ipvs/ip_vs_service.c:47]: we are trying to add 100k+ snat svc, 8-bit is not enough

File: src/ipvs/ip_vs_service.c [hunk 3]
```diff
@@ -36,21 +37,46 @@
...
 #define DP_VS_SVC_TAB_SIZE (1 << DP_VS_SVC_TAB_BITS)
 #define DP_VS_SVC_TAB_MASK (DP_VS_SVC_TAB_SIZE - 1)
 
 static struct list_head dp_vs_svc_table[DPVS_MAX_LCORE][DP_VS_SVC_TAB_SIZE];
 
 static struct list_head dp_vs_svc_fwm_table[DPVS_MAX_LCORE][DP_VS_SVC_TAB_SIZE];
 
-static struct list_head dp_vs_svc_match_list[DPVS_MAX_LCORE];
+/* lock for master lcore */
+static rte_rwlock_t dp_vs_svc_match_lock[DPVS_MAX_LCORE];
```
[the remark this answers is not in the document]
Comment (donghaobo) [reply src/ipvs/ip_vs_service.c:56]: this lock is for svc match ACL build thread.  and it is per-core, will be used only at svc update. dataplan will not use it. 
i know the svc operation is processed in slave workers, this lock will slow down the svc insert/delete a little. 
but this lock is  hard to remove. we are planing to modify the control plane with RCU, and that is another project.

File: src/ipvs/ip_vs_service_match_acl.c (hunk 1 above)
Comment (donghaobo) [reply src/ipvs/ip_vs_service_match_acl.c:877]: 10w dpdk ACL rules compiling will take about 10 seconds, this should not do in the master. a new thread is very needed.

**Comment by SimonCZW** [comment]
How much performance has been improved with this solution? @donghaobo

**Comment by donghaobo** [comment]
> How much performance has been improved with this solution? @donghaobo

this commit is used with 10k+ SNAT rules。it's trying to reach the same performance as only one SNAT rule . if there's only a default SNAT rule, this commit does nothing help.
if you really run dpvs with 10k+ SNAT rules, you will know how bad the performance is .

File: src/netif.c [hunk 4]
```diff
@@ -1120,7 +1120,8 @@ int dpvs_wait_lcores(void)
     do {
         sched_yield();
         for (i = 0; i < DPVS_MAX_LCORE; i++) {
-            if (is_lcore_id_fwd(i) && loop[i] == lcore_stats[i].lcore_loop) {
```
Comment (donghaobo) [inline src/netif.c]: pr #669  need this change

**Comment by yifan1227** [comment]
May I ask in what circumstances 10k+ SNAT rules are needed?  How many public IPs do you configure for each SNAT director?

**Pull request closed without merging**